ffi/arcreader.lua

```diff
@@ -0,0 +1,178 @@
+--[[--
+@module ffi.arcreader
+]]
+
+local ffi = require "ffi"
+local libarchive = ffi.loadlib("archive", "13")
+require "ffi/libarchive_h"
+require "ffi/posix_h"
+
+local ArcReader = {}
+
+local ENTRY_MODE = {
+    [libarchive.AE_IFREG]  = "file",
+    [libarchive.AE_IFLNK]  = "link",
+    [libarchive.AE_IFSOCK] = "socket",
+    [libarchive.AE_IFCHR]  = "char device",
+    [libarchive.AE_IFBLK]  = "block device",
+    [libarchive.AE_IFDIR]  = "directory",
+    [libarchive.AE_IFIFO]  = "named pipe",
+}
+
+local function archive_error_string(archive)
+    local err = libarchive.archive_error_string(archive)
+    return err ~= nil and ffi.string(err) or nil
+end
+
+function ArcReader:new()
+    local o = {}
+    setmetatable(o, self)
+    self.__index = self
+    o.archive_entry = ffi.new("struct archive_entry *[1]")
+    o.entries = {}
+    o.size = 0
+    o.err = nil
+    return o
+end
+
+function ArcReader:open(filepath)
+    self.err = nil
+    if not filepath then
+        return
+    end
+    self.archive = ffi.gc(libarchive.archive_read_new(), libarchive.archive_free)
+    libarchive.archive_read_support_format_all(self.archive)
+    libarchive.archive_read_support_filter_all(self.archive)
+    if libarchive.archive_read_open_filename(self.archive, filepath, 10240) ~= libarchive.ARCHIVE_OK then
+        self.err = archive_error_string(self.archive)
+        print("ArcReader: archive_read_open_filename failed:", self.err)
+        libarchive.archive_free(self.archive)
+        ffi.gc(self.archive, nil)
+        self.archive = nil
+        return
+    end
+    self.filepath = filepath
+    self.index = 0
+    return true
+end
+
+function ArcReader:next()
+    self.err = nil
+    if self.archive == nil then
+        return
+    end
+    -- We can't consume the data twice: tweak the index
+    -- to trigger a reset in case of subsequent attempt.
+    self.index = math.floor(self.index) + 0.1
+    local err = libarchive.archive_read_next_header(self.archive, self.archive_entry)
+    if err ~= libarchive.ARCHIVE_OK then
+        if err ~= libarchive.ARCHIVE_EOF then
+            self.err = archive_error_string(self.archive)
+            print("ArcReader: archive_read_next_header failed:", self.err)
+        end
+        return
+    end
+    local entry
+    self.index = math.floor(self.index) + 1
+    if self.index > self.size then
+        local archive_entry = self.archive_entry[0]
+        entry = {
+            path = ffi.string(libarchive.archive_entry_pathname(archive_entry)),
+            mode = ENTRY_MODE[libarchive.archive_entry_filetype(archive_entry)] or "other",
+            size = libarchive.archive_entry_size(archive_entry),
```

Comment:
I don't recall the libarchive types (on mobile right now) but If it's a 64bit wide type on 32bit arches, it'll need to be truncated (e.g., tonumber).

Or, maybe better given that it is sort of critical not to screw this one up, keep all the maths on it in C ;). (e.g., keep self.size as a C type).

Reply:
Nevermind me, clearly haven't gotten enough sleep ;).

Confused the index size with this entry size field ;p.